src/lib.rs

```diff
@@ -0,0 +1,780 @@
+    /// to which direction the stack grows and different clients may
+    /// have different preferences.
+    fn multi_spillslot_named_by_last_slot(&self) -> bool {
+        false
+    }
+}
+
+/// A position before or after an instruction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(u8)]
+pub enum InstPosition {
+    Before = 0,
+    After = 1,
+}
+
+/// A program point: a single point before or after a given instruction.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ProgPoint {
+    pub inst: Inst,
+    pub pos: InstPosition,
+}
+
+impl ProgPoint {
+    pub fn before(inst: Inst) -> Self {
+        Self {
+            inst,
+            pos: InstPosition::Before,
+        }
+    }
+
+    pub fn after(inst: Inst) -> Self {
+        Self {
+            inst,
+            pos: InstPosition::After,
+        }
+    }
+
+    pub fn next(self) -> ProgPoint {
+        match self.pos {
+            InstPosition::Before => ProgPoint {
+                inst: self.inst,
+                pos: InstPosition::After,
+            },
+            InstPosition::After => ProgPoint {
+                inst: self.inst.next(),
+                pos: InstPosition::Before,
+            },
+        }
+    }
+
+    pub fn prev(self) -> ProgPoint {
+        match self.pos {
+            InstPosition::Before => ProgPoint {
+                inst: self.inst.prev(),
+                pos: InstPosition::After,
+            },
+            InstPosition::After => ProgPoint {
+                inst: self.inst,
+                pos: InstPosition::Before,
+            },
+        }
+    }
+
+    pub fn to_index(self) -> u32 {
+        debug_assert!(self.inst.index() <= ((1 << 31) - 1));
+        ((self.inst.index() as u32) << 1) | (self.pos as u8 as u32)
+    }
+
+    pub fn from_index(index: u32) -> Self {
+        let inst = Inst::new((index >> 1) as usize);
+        let pos = match index & 1 {
+            0 => InstPosition::Before,
+            1 => InstPosition::After,
+            _ => unreachable!(),
+        };
+        Self { inst, pos }
+    }
+}
+
+/// An instruction to insert into the program to perform some data movement.
+#[derive(Clone, Debug)]
+pub enum Edit {
+    /// Move one allocation to another. Each allocation may be a
+    /// register or a stack slot (spillslot).
+    Move { from: Allocation, to: Allocation },
+    /// Define blockparams' locations. Note that this is not typically
+    /// turned into machine code, but can be useful metadata (e.g. for
+    /// the checker).
+    BlockParams {
+        vregs: Vec<VReg>,
+        allocs: Vec<Allocation>,
+    },
+}
+
+/// A machine envrionment tells the register allocator which registers
+/// are available to allocate and what register may be used as a
+/// scratch register for each class, and some other miscellaneous info
+/// as well.
+#[derive(Clone, Debug)]
+pub struct MachineEnv {
+    regs: Vec<PReg>,
+    regs_by_class: Vec<Vec<PReg>>,
```

Comment:
I looked briefly at this but it seems to pull in some other deps (e.g. a derive macro, and through that, the whole `syn` crate and deps); I'd prefer to keep the dependency set very small at the bottom of the dep-graph, especially if the benefit otherwise is just a small ergonomics thing. Right now it's just `fxhash`, `smallvec`, `log`.